lib/dns/dns.go

```diff
@@ -288,37 +289,72 @@ func (s *Server) HandleForwarding(w mdns.ResponseWriter, r *mdns.Msg) (bool, err
 	// Use request ID for "random" nameserver selection.
 	nsid := int(r.Id) % len(s.Nameservers)
 
-Redo:
-	nameserver := s.Nameservers[nsid]
-	if i := strings.Index(nameserver, ":"); i < 0 {
-		nameserver += ":53"
-	}
+	for try := 0; try < len(s.Nameservers); try++ {
+		nsid = (nsid + 1) % len(s.Nameservers)
 
-	if tcp {
-		m, _, err = s.tcpclient.Exchange(r, nameserver)
-	} else {
-		m, _, err = s.udpclient.Exchange(r, nameserver)
-	}
-	if err != nil {
-		// Seen an error, this can only mean, "server not reached", try again but only if we have not exausted our nameservers.
-		if try < len(s.Nameservers) {
-			try++
-			nsid = (nsid + 1) % len(s.Nameservers)
-			goto Redo
+		nameserver := s.Nameservers[nsid]
+		if i := strings.Index(nameserver, ":"); i < 0 {
+			nameserver += ":53"
+		}
+
+		log.Debugf("Forwarding request to %s", nameserver)
+
+		if tcp {
+			m, _, err = s.tcpclient.Exchange(r, nameserver)
+		} else {
+			m, _, err = s.udpclient.Exchange(r, nameserver)
+		}
+
+		if err != nil || m == nil {
+			// Seen an error, this can only mean, "server not reached", or similar
+			log.Errorf("Failure to forward request to %s: %q", nameserver, err)
+			continue
 		}
 
-		log.Errorf("Failure to forward request: %q", err)
+		if m.Authoritative {
+			// don't even check whether it was a successful return. If it's authoritative, then that's
+			// the answer
+			log.Debugf("Authoritative response")
+			authoritativeResp = m
+			break
+		}
+
+		// non-authoritative processing
+		// nothing clever:
+		// * if a success response, stash it
+		// * if not a success response, keep it if no previous stashed response
+		if m.Rcode == mdns.RcodeSuccess {
+			log.Debugf("Non-authoritative success")
+			nonAuthoritativeResp = m
```

Comment:
I'm not sure this leads to the intended behavior.

If we have 3 nameservers, the following could occur:

1. The first returns a non-authoritative non-successful response. We store it as we have no previous stashed response.
2. The second returns a non-authoritative successful response. We (correctly) store it as its "better" the non-successful we currently have stashed.
3. The third returns a non-authoritative successful response. We store it even though we already have a successful response stashed, overwriting the response from the second nameserver.

In this case, I would naively expect the non-authoritative response from the second nameserver to take precedence over the non-authoritative response from the third nameserver.

I'm not sure what the correct behavior is.

Reply:
@zjs We "randomize" the query order (not that I'm sure why) so I went with simplest code given there's no explicit predictability anyway.